client/src/templates/Challenges/odin/show.tsx

```diff
@@ -166,33 +176,45 @@ class ShowOdin extends Component<ShowOdinProps, ShowOdinState> {
     }
   }
 
-  handleSubmit(
-    solution: number,
-    openCompletionModal: () => void,
-    assignments: string[]
-  ) {
-    const hasAssignments = assignments.length > 0;
-    const completed = this.state.allAssignmentsCompleted;
-    const isCorrect = solution - 1 === this.state.selectedOption;
-
-    if (isCorrect) {
-      this.setState({
-        isWrongAnswer: false
-      });
-      if (!hasAssignments || completed) openCompletionModal();
-    } else {
-      this.setState({
-        isWrongAnswer: true
-      });
+  handleSubmit = () => {
+    const {
+      data: {
+        challengeNode: {
+          challenge: { questions }
+        }
+      },
+      openCompletionModal
+    } = this.props;
+
+    // subract 1 because the solutions are 1-indexed
+    const mcqSolutions = questions.map(question => question.solution - 1);
+
+    this.setState({
+      submittedMcqAnswers: this.state.selectedMcqOptions,
+      showFeedback: true
+    });
+
+    const allMcqAnswersCorrect = arraysAreEqual(
+      mcqSolutions,
+      this.state.selectedMcqOptions
+    );
+
+    if (this.state.allAssignmentsCompleted && allMcqAnswersCorrect) {
+      openCompletionModal();
     }
-  }
+  };
 
-  handleOptionChange = (
-    changeEvent: React.ChangeEvent<HTMLInputElement>
+  handleMcqOptionChange = (
+    questionIndex: number,
+    answerIndex: number
   ): void => {
+    const { selectedMcqOptions } = this.state;
+
+    const newSelectedMcqOptions = [...selectedMcqOptions];
+    newSelectedMcqOptions[questionIndex] = answerIndex;
+
     this.setState({
-      isWrongAnswer: false,
-      selectedOption: parseInt(changeEvent.target.value, 10)
+      selectedMcqOptions: newSelectedMcqOptions
```

Comment:
The UI doesn't support this suggestion, but:

```ts
    this.setState(state => ({
      selectedMcqOptions: state.selectedMcqOptions.map((option, index) =>
        index === questionIndex ? answerIndex : option
      )
    }));
```
This is a bit neater, imo, but the main thing is that any `setState` that depends on existing state should use the callback. Otherwise state updates can get lost when React batches them.

Reply:
✅